web/modules/custom/nfafmis/nfafmis.module

```diff
@@ -798,3 +759,199 @@ function nfafmis_form_ajax_validate(array &$form, FormStateInterface $form_state
+      $range = $area->field_central_forest_reserve->entity->management_unit->entity;
+      // Range Users are granted 'view' permission to nodes in their range.
+      // Remove duplicates by setting explicit array key.
+      $grants[$range->id()] = [
+        'realm' => nfafmis_range_access_realm($range),
+        'gid' => $range->id(),
+        'grant_view' => 1,
+        'grant_update' => 0,
+        'grant_delete' => 0,
+        'priority' => 0,
+      ];
+      // Range Power Users are granted all permissions to nodes in their range.
+      $grants[$range->id() + 10000] = [
+        'realm' => nfafmis_range_access_realm($range, TRUE),
+        'gid' => $range->id(),
+        'grant_view' => 1,
+        'grant_update' => 1,
+        'grant_delete' => 1,
+      ];
+    }
+  }
+  else {
+    if ($node->bundle() == 'offer_license') {
+      // Retrieve the CFR and Range (Management Unit) that the area is in.
+      $range = $node->field_central_forest_reserve->entity->management_unit->entity;
+    }
+    elseif ($node->bundle() == 'sub_area') {
+      // Retrieve the CFR and Range (Management Unit) that the sub-area's parent
+      // area is in.
+      if ($node->field_areas_id->entity && $cfr = $node->field_areas_id->entity->field_central_forest_reserve->entity) {
+        $range = $cfr->management_unit->entity;
+      }
+      else {
+        // @todo flag sub areas with no assigned area and areas with no
+        // assigned CFRs.
+        $stop = TRUE;
+      }
+    }
+
+    if ($range) {
+      // Range Users are granted 'view' permission to nodes in their range.
+      $grants[] = [
+        'realm' => nfafmis_range_access_realm($range),
+        'gid' => $range->id(),
+        'grant_view' => 1,
+        'grant_update' => 0,
+        'grant_delete' => 0,
+      ];
+      // Range Power Users are granted all permissions to nodes in their range.
+      $grants[] = [
+        'realm' => nfafmis_range_access_realm($range, TRUE),
+        'gid' => $range->id(),
+        'grant_view' => 1,
+        'grant_update' => 1,
+        'grant_delete' => 1,
+      ];
+    }
+  }
+
+  // HQ Users are granted view permissions to all nodes.
+  $grants[] = [
+    'realm' => 'nfafmis_view',
+    'gid' => 1,
+    'grant_view' => 1,
+    'grant_update' => 0,
+    'grant_delete' => 0,
+  ];
+
+  // HQ Power Users are granted all permissions to all nodes.
+  $grants[] = [
+    'realm' => 'nfafmis_all',
+    'gid' => 1,
+    'grant_view' => 1,
+    'grant_update' => 1,
+    'grant_delete' => 1,
+  ];
+
+  return $grants;
+}
+
+/**
+ * Implements hook_node_grants().
+ * Return the grants (node access permissions) that this user account has.
+ */
+function nfafmis_node_grants(AccountInterface $account, $op) {
+  $grants = [];
+  $roles = $account->getRoles();
+
+  // If the user is a Range user, retrieve the range that the user has been
+  // assigned to.
+  if (in_array('range_user', $roles) || in_array('range_power_user', $roles)) {
```

Comment:
I'm not very fond of this technique, I rather create a permission than checking user roles.

Reply:
I added two new permissions in lieu of checking user roles